### ISSUE TITLE
refactor: retire flock's software field.py — compute the PIOP in native binary_field_ghash

### ISSUE THREAD

**Issue by chokobole** [fractalyze/flock-zorch#45]
### Pre-submission checklist

- [x] I have searched the existing issues and this is not a duplicate.

### Context / problem

flock-zorch carries its own software GF(2^128) field, `python/flock_zorch/field.py` — a hand-rolled GHASH-basis multiply (uint64-lane CLMUL emulation) plus XOR add/sum, over a `uint64[..., 2] = [lo, hi]` representation. Its `mul` is threaded as a `mul=` parameter through ~63 signatures so a faster backend could be swapped in (the now-retired `field_clmad.py` FFI, PR #44). The milestone's original design fact — "Field stays flock's GHASH (field.py); zk_dtypes is tower-basis only, no Tower↔GHASH bridge" — was true then but is now **obsolete**: #27 landed `binary_field_ghash` as a first-class non-tower zk_dtypes dtype.

### Proposed solution

Retire `field.py` by computing flock's PIOP arithmetic directly in the native `binary_field_ghash` dtype, and drop the `mul=` seam. Verified this session (CPU+GPU):
- add → `a + b`, sum → `jnp.sum`, mul → `a * b` are **byte-identical** to `field.{add,sum,mul}` on both backends.
- flock's PIOP needs **no field inverse** (field.py has none; the `inv` in `_hostfield.py`/`gf8.py` are separate modules).
- the dtype's `*` is the GPU-accelerated multiply, so `field_clmad.py` is subsumed.

Migrate incrementally, module by module — each keeping its `*_oracle_test` byte-green on CPU and GPU — then delete `field.py` once nothing computes with it.

### Alternatives considered

- **Keep `field.py` + a `GhashFieldOps` seam** (the milestone's original plan) — tried and dropped in PR #38; added indirection without removing the software field.
- **Bridge-per-op** (keep uint64[n,2], convert to ghash for each multiply) — rejected: per-op bridging adds overhead and keeps the software field alive. The clean end-state computes in the dtype and serializes to uint64 only at gate edges.

### Why this approach

Completes the zorch-native reuse (flock stops maintaining a hand-rolled software field zk_dtypes now provides), removes ~63 `mul=` parameters, and unlocks the GPU-accelerated multiply everywhere without an FFI. Equivalence is proven byte-for-byte, and PR #43 (ring-switch) already shows a full module consuming the dtype byte-green.

### Plan

- [ ] De-risk: migrate `sumcheck.py` (shared kernel) to compute in `binary_field_ghash`, keeping `sumcheck_oracle_test` byte-green CPU+GPU. (This issue.)
- [ ] Sweep the remaining ~17 modules (zerocheck, lincheck, basefold, prover, chain, gf8, fri, …), one per PR, each keeping its oracle gate byte-green.
- [ ] Drop `mul=` parameters as each module migrates.
- [ ] Keep flock's `uint64[n,2]` serde only at golden-gate boundaries, bridged via **uint32 lanes** — the direct `uint64 ↔ binary_field_ghash` bitcast silently miscompiles on CPU PJRT (a fractalyze/xla BitcastConvertType bug); `uint32 ↔ ghash` is correct on both backends.
- [ ] Delete `field.py` (or reduce to `LOG_PACKING` + serde helpers) once no module imports `field.{mul,add,sum}`.

### Acceptance criteria

- No module imports `field.mul` / `field.add` / `field.sum`.
- `field.py` deleted, or reduced to serialization helpers only.
- Every `*_oracle_test` byte-green on CPU and GPU.
- `field_clmad.py` gone (done — PR #44).

### References

- #27 — `binary_field_ghash` dtype (enabling dependency)
- PR #43 — ring-switch consumes the dtype (full-module byte-green precedent)
- PR #44 — `field_clmad.py` retired
- PR #38 — the dropped `GhashFieldOps` seam

### Relevant knowledge

**uint64 ↔ binary_field_ghash bitcast silently miscompiles on CPU PJRT — bridge via uint32.** `lax.bitcast_convert_type` between `uint64[..., 2]` and `binary_field_ghash` returns wrong bytes (no error) on CPU; `uint32[..., 4] ↔ ghash` is correct on both backends, and uint32 is the lane width zorch's kernels use natively. Any flock↔dtype boundary must route through uint32:

```python
def to_ghash(u64):   # uint64[...,2] -> ghash[...]
    u32 = lax.bitcast_convert_type(u64, jnp.uint32).reshape(*u64.shape[:-1], 4)
    return lax.bitcast_convert_type(u32, jnp.binary_field_ghash)
def from_ghash(g):   # ghash[...] -> uint64[...,2]
    u32 = lax.bitcast_convert_type(g, jnp.uint32)
    return lax.bitcast_convert_type(u32.reshape(*u32.shape[:-1], 2, 2), jnp.uint64)
```

Depends on: #27

**Comment by chokobole** [comment]
### Session 2026-07-07 progress — field.py endgame Phase 1 complete

- **Done:** **Phase 1** — every module's device GF(2^128) arithmetic migrated off flock's software `field.mul/add/sum` to the native `binary_field_ghash` dtype (`*` / `+` / `jnp.sum`), across `sumcheck`, `fri`, `gf8`/`_gf8_device`, `keccak_lincheck` (+`keccak3_lincheck`), `chain`, `basefold`, `lincheck`, `ligerito`, `prover`, `_zerocheck_fold`. Branch **`refactor/field-endgame`** (5 commits, pushed, off `main`). **`bazel //python:all` 21/21 on CPU + GPU `zerocheck`/`e2e` full-prove byte-match.** The `uint64[..,2]↔binary_field_ghash` bitcast bridge is centralized in `field.to_ghash` / `field.from_ghash` (uint32-routed). `field.mul/add/sum` are no longer called for **compute** anywhere.
- **Next session:** do **(2)** — the fully edge-only ghash-native I/O flip **+ delete `field.py`** (plan below).
- **Blocker:** none.

---

## Phase 2 plan (option 2 — arrays are `binary_field_ghash` end-to-end, bridge only at edges, `field.py` deleted)

1. **Strip every `mul=` parameter + call-site arg.** ~19 `mul=field.mul` defaults + `pcs.py:42` `mul: Callable = field.mul`. Every function already ignores `mul` (native ops), so this is mechanical and stays gate-green. Find them: `grep -rn 'mul=' python/flock_zorch/*.py`.
2. **Flip module I/O to the dtype.** Remove each module's internal `field.to_ghash`/`from_ghash` so functions take/return `binary_field_ghash`; the bridge then survives ONLY at the two real edges — the golden-fixture readers in `testing/*_oracle_test.py` and the challenger's F128 byte serde (`challenger.py` `observe_f128*`/`sample_f128*`). This is **coupled** (flipping a module's I/O breaks its callers), so do it as one coordinated pass in dependency order and gate at the end.
3. **Rehome `basefold.verify`'s host `np.uint64` replay** (~lines 297–316, the only remaining real `field.add`/`mul` code) → host-int `_hostfield` (`hf.mul`) or the ghash bridge.
4. **Move serde out of `field.py`** — `to_ghash`/`from_ghash`, `_to_int`/`_to_lohi`, `LOG_PACKING` → `_hostfield.py` (or a new `_serde.py`); update imports; **delete `field.py`**.

## Gotchas (verified this session — will bite again)
- **`uint64↔binary_field_ghash` bitcast silently miscompiles on CPU PJRT** (wrong bytes, no error) — always route through **uint32 lanes** (that is exactly what `field.to_ghash/from_ghash` do). Never bitcast uint64↔ghash directly.
- **A long *sequential* chain of native ghash `*` compiles pathologically slowly on CPU** (minutes → test timeout). The 127-step Fermat inverse `_zerocheck_fold._batch_inv` had to be a `lax.fori_loop`, not a Python unroll. Vectorized/parallel multiplies over an array axis are fine — only the unrolled sequential chain hurts.
- **0/1 selection products are integer, not field muls** — e.g. `mat_dense[..]*eq[..]` where `mat_dense∈{0,1}` selects; keep `*` on uint64 and bridge only the following XOR-reduce (`jnp.sum` on ghash). See `lincheck._mat_fold`, `_zerocheck_fold._fold_at_z_dev`.
- `sumcheck.ONE` is the ghash scalar one; reuse it (or `field.to_ghash([1,0])`).

## Verify recipe
```
cd ~/Workspace/flock-zorch2
bazel test //python:all --test_output=errors            # CPU, must stay 21/21
# GPU spot-check (whole prove path):
XLA_PYTHON_CLIENT_PREALLOCATE=false JAX_COMPILATION_CACHE_DIR="" JAX_PLATFORMS=cuda JAX_ENABLE_X64=1 \
  PYTHONPATH=python:/home/ryan/Workspace/envs/flock-zorch2/zorch \
  .venv/bin/python python/flock_zorch/testing/e2e_oracle_test.py
```
Devenv: zorch override at `envs/flock-zorch2/zorch` on branch `flock-byte-fiat-shamir @ f9b6593` (== main's `git_override` pin). Goldens regenerate via `bash scripts/dump_goldens.sh`.

## Reference files
- `python/flock_zorch/field.py` — bridge + serde (to be emptied/deleted)
- `python/flock_zorch/sumcheck.py`, `.../fri.py` — the migration templates
- `python/flock_zorch/_hostfield.py` — rehome target for serde/host-int
- `python/flock_zorch/challenger.py` — the F128-byte transcript edge (a bridge boundary)
- `python/flock_zorch/testing/*_oracle_test.py` — the golden-fixture edge (the other bridge boundary)

## Done criteria
1. No module imports/calls `field.mul` / `field.add` / `field.sum`; no `mul=` params remain.
2. Modules pass/return `binary_field_ghash`; the uint64 bitcast survives only at the gate + challenger edges.
3. `field.py` deleted (serde rehomed).
4. `bazel //python:all` 21/21 (CPU) + GPU `e2e`/`zerocheck` byte-match.

## Out of scope
- `ring_switch`/`ligerito` dtype paths already landed by #43/#39 — don't re-touch.
- Host-int Fermat/Lagrange (`_to_int`/`_to_lohi`/`hf.mul`, `_hostfield`) stays host — it is NOT device dtype arithmetic.
- Whether to PR Phase 1 separately vs bundle with Phase 2: branch is pushed unmerged; decide at PR time. (The `#46` sumcheck de-risk PR is folded into this branch via cherry-pick — close it as superseded, or let this branch's PR supersede it.)

**Comment by chokobole** [comment]
### Post-merge note 2026-07-07 — validation scope + one loose end

Merged as `bc748f5` (PR #55), rebased onto `main` (absorbed the in-tree ligerito retirement #48, the basefold multi-epoch anchor #53, the merkle octopus dedup #54, and the zorch pin bump to `7f6b3db`).

- **Done:** the whole endgame — device arithmetic on native `binary_field_ghash`, the `mul=` seam removed, `field.py` reduced to the uint64-lane↔dtype serde bridge, software `field.{mul,add,sum}` + `field_oracle_test`/golden/`dump_field_mul.rs` retired. `basefold_verify`'s reference target now computed host-side in `_hostfield` (independent of the dtype under test).
- **Validated:** `bazel test //python:all` **20/20** (CPU) + `e2e_oracle_test.py` full-prove **byte-match vs flock** on GPU (CUDA).
- **Blocker:** none.

**Loose end — heavy hash-circuit GPU oracle gates were NOT run** (keccak/sha2/blake3 — venv-only, excluded from the bazel/CI CPU set, so unverified anywhere on this branch). Edits to the `keccak_lincheck`/`keccak3_lincheck` walkers were trivial dead-`mul=` removals and the rebased `zorch_ligerito` path is CI-covered on `main`, so risk is low — but to close the milestone invariant ("every `*_oracle_test` byte-green on GPU") they should be run:

```bash
cd ~/Workspace/flock-zorch2
export JAX_PLATFORMS=cuda XLA_PYTHON_CLIENT_PREALLOCATE=false
export PYTHONPATH="python:$(scripts/zorch_pythonpath.sh)"
scripts/dump_goldens.sh all      # dumps the heavy keccak/sha2/blake3 goldens (slow)
.venv/bin/python python/flock_zorch/testing/keccak_oracle_test.py
.venv/bin/python python/flock_zorch/testing/sha2_ligerito_oracle_test.py
.venv/bin/python python/flock_zorch/testing/blake3_ligerito_oracle_test.py
```

Each should print a byte-match PASS. If one fails, the regression is in the `mul=` strip of the keccak/lincheck walkers or the rebased `zorch_ligerito` open path.

**Separate pre-existing cleanup (not #45):** `optim/clmad/`'s README + FFI still name `field.mul` / `field_clmad`, both dead since the #44 FFI retirement. The clmad GPU multiply is now the `binary_field_ghash` dtype's concern (emitted by the zkx compiler); the directory wants its own docs pass.

**Comment by chokobole** [comment]
### Loose end closed 2026-07-08 — heavy hash-circuit GPU oracle gates are byte-green

The one unverified item from the [post-merge note](https://github.com/fractalyze/flock-zorch/issues/45#issuecomment-4909671647) — the venv-only keccak/sha2/blake3 gates excluded from the bazel/CI CPU set — has now been run on GPU. **All three byte-match unmodified flock**, so the milestone invariant "every `*_oracle_test` byte-green on GPU" is fully satisfied for the field.py endgame.

| Gate | Circuit | Result |
|------|---------|--------|
| `keccak_oracle_test` | full BaseFold prove, m=19 (commit + zerocheck + walker lincheck + batched open) | **PASS** |
| `sha2_ligerito_oracle_test` | full Ligerito prove (`R1csProofLigerito`), m=22 | **PASS** |
| `blake3_ligerito_oracle_test` | full Ligerito prove (`R1csProofLigerito`), m=22 | **PASS** |

Every sub-check (commit root, zerocheck rounds, lincheck, ring-switch opens, and the full Ligerito/BaseFold transcript incl. recursive roots, grinding nonces, merkle proofs) printed PASS.

**Environment:** branch `refactor/field-endgame @ 6465a92` (the code merged as PR #55), zorch override `@ 7f6b3db`, RTX 5090 / CUDA. Goldens regenerated fresh from the pinned `third_party/flock` submodule via `cargo run --release --example dump_{keccak,sha2_ligerito,blake3_ligerito}`. Ran with `JAX_PLATFORMS=cuda JAX_ENABLE_X64=1 XLA_PYTHON_CLIENT_PREALLOCATE=false JAX_COMPILATION_CACHE_DIR=""` (empty cache dir — correctness gate against a self-built jaxlib).

Verified the gates genuinely exercise the migrated path: `field.py` no longer defines `mul`/`add`/`sum`, and the tests import `field` only for the uint32-lane serde bridge — the arithmetic runs native `binary_field_ghash`.

**Nit (not blocking):** the three test headers print a hardcoded `| mul software` label (e.g. `keccak_oracle_test.py:148`) — vestigial text from before the migration; there is no software mul anymore. Worth dropping on the next docs/test-cosmetics pass.

**Still open (separate, not #45):** the `optim/clmad/` README + FFI still name `field.mul` / `field_clmad`, dead since the #44 FFI retirement — flagged in the post-merge note, wants its own cleanup.
